include/ur_api.h

```diff
@@ -1051,18 +1103,23 @@ typedef enum ur_device_partition_t {
 ///         + `NULL == hDevice`
 ///     - ::UR_RESULT_ERROR_INVALID_NULL_POINTER
 ///         + `NULL == pProperties`
+///     - ::UR_RESULT_ERROR_INVALID_ENUMERATION
+///         + `::UR_DEVICE_PARTITION_BY_CSLICE < pProperties->type`
 ///     - ::UR_RESULT_ERROR_DEVICE_PARTITION_FAILED
 ///     - ::UR_RESULT_ERROR_INVALID_DEVICE_PARTITION_COUNT
+///     - ::UR_RESULT_ERROR_INVALID_SIZE
+///         + `DescCount == 0`
 UR_APIEXPORT ur_result_t UR_APICALL
 urDevicePartition(
-    ur_device_handle_t hDevice,                        ///< [in] handle of the device to partition.
-    const ur_device_partition_property_t *pProperties, ///< [in] null-terminated array of <$_device_partition_t enum, value> pairs.
-    uint32_t NumDevices,                               ///< [in] the number of sub-devices.
-    ur_device_handle_t *phSubDevices,                  ///< [out][optional][range(0, NumDevices)] array of handle of devices.
-                                                       ///< If NumDevices is less than the number of sub-devices available, then
-                                                       ///< the function shall only retrieve that number of sub-devices.
-    uint32_t *pNumDevicesRet                           ///< [out][optional] pointer to the number of sub-devices the device can be
-                                                       ///< partitioned into according to the partitioning property.
+    ur_device_handle_t hDevice,                    ///< [in] handle of the device to partition.
+    const ur_device_partition_desc_t *pProperties, ///< [in] Array of partition descriptors.
+    size_t DescCount,                              ///< [in] Number of descriptors pointed to by `pProperties`.
```

Comment:
I think the array and count should be in the properties struct and rather than having each `ur_device_partition_desc_t` have `stype` and `pNext` don't call it a desc. Then we both have extensibility and a simple array query for the properties.

```cpp
struct ur_device_partition_property_t {
    ur_device_partition_t type;
    ur_device_partition_value_t value;
};

struct ur_device_partition_properties_t {
    ur_structure_type_t stype;
    void* pNext;
    const ur_device_partition_property_t *pProperties;
    size_t DescCount;
};

UR_APIEXPORT ur_result_t UR_APICALL urDevicePartition(
    ur_device_handle_t hDevice,
    const ur_device_partition_properties_t *pProperties,
    uint32_t NumDevices,
    ur_device_handle_t *phSubDevices,
    uint32_t *pNumDevicesRet);
```

Reply:
I've implemented your suggestion 👍